ezsystems/ezpublish-kernel
# EZP-27743: Criterion should implement CriterionInterface

| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-27743](https://jira.ez.no/browse/EZP-27743)
| **Bug/Improvement**| Improvement
| **New feature**    | no
| **Target version** | `master`
| **BC breaks**      | yes
| **Tests pass**     | yes
| **Doc needed**     | yes

This PR is one of the series that aims at cleaning up the code related to Criterions.
Current issues:
* The abstract class `eZ\Publish\API\Repository\Values\Content\Query\Criterion\LogicalOperator` doesn't implement `eZ\Publish\API\Repository\Values\Content\Query\CriterionInterface`, but extends `eZ\Publish\API\Repository\Values\Content\Query\Criterion` abstract class instead. This makes `CriterionInterface` not used almost at all, and it leads to `CriterionInterface|LogicalOperator` return type hint combo when it would make sense to have only CriterionInterface instead (see https://github.com/ezsystems/ezpublish-kernel/pull/2035).
* All classes extending abstract class `Criterion` implement `CriterionInterface`, even though `Criterion` does not. However, `Criterion` abstract class uses `getSpecifications` method that is present in `CriterionInterface`, and not in `Criterion`.
* The abstract class `LogicalOperator` extends `Criterion` abstract class, but doesn't use any code from there:
   * `LogicalOperator` doesn't use the constructor from the `Criterion` class, it overwrites it completely: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Values/Content/Query/Criterion/LogicalOperator.php#L34 
   * As a result, `LogicalOperator` doesn't use the `getValueTypeCheckCallback` private function from `Criterion` class: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Values/Content/Query/Criterion.php#L123
   * `LogicalOperator` inherits `createFromQueryBuilder` function from `Criterion` class, which looks plain wrong in `LogicalOperator`'s scope (it passes three arguments to the constructor when `LogicalOperator` has only one and it is different one): https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Values/Content/Query/Criterion.php#L149
   * `LogicalOperator` declares its own attribute and I don't think the attributes from `Criterion` class make much sense in `LogicalOperator`'s scope: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Values/Content/Query/Criterion/LogicalOperator.php#L2

What this PR fixes:
* The abstract class `Criterion` now implements interface `CriterionInterface`.
* The `createFromQueryBuilder` and `getSpecifications` methods were moved from `CriterionInterface` to `Criterion` abstract class (the former was already there). This was done because neither of these methods is supposed to be implemented by `LogicalOperator`, which will implement this interface directly in the future.
* The method `createFromQueryBuilder` mentioned in the previous point was deprecated. This was done because I couldn't find any code that depends on it and under the hood it just calls the constructor. I think that one way to create an object is enough unless another way is specifically required.
* Since the abstract class `Criterion` now implements interface `CriterionInterface`, all classes that extend the `Criterion` no longer need to directly implement this interface, so `implements CriterionInterface` was removed from them.
* Abstract class `LogicalOperator` now has the new method `getSpecifications` which throws `eZ\Publish\API\Repository\Exceptions\NotImplementedException` and is immediately deprecated. This is a temporary measure until `LogicalOperator` will no longer extend `Criterion`.

This PR shouldn't break any code that depends on changed classes.

What is planned to be done:
* Remove `extends Criterion` from `LogicalOperator` and replace it with `implements CriterionInterface`. This will be done as part of 8.0 since it is a major BC break (types are changed).
* Check the codebase for all usages of `Criterion`, `CriterionInterface` and `LogicalOperator`. Change the usages where appropriate (`Criterion` should be used when singular Criterion should be expected, `LogicalOperator` when the operator should be expected and `CriterionInterface` when either can be expected). This should fix all `CriterionInterface|LogicalOperator` return type hint combos (among other things).

**TODO**:
- [x] Implement tests (skipped).
